### PR TITLE
Usage example using IRIS dataset to tf.data.experimental.make_csv_dataset()

### PR DESCRIPTION
**Pull request by boronhub**
Add usage example using IRIS dataset to tf.data.experimental.make_csv_dataset()

GCI 2019

### CODE REVIEW

File: tensorflow/python/data/experimental/ops/readers.py [hunk 1]
```diff
@@ -420,6 +420,28 @@ def make_csv_dataset_v2(
 
   Raises:
     ValueError: If any of the arguments is malformed.
+  Usage Example:
+  
+  Using IRIS dataset to show how to convert .csv file into a dataset.
+  
+  ```python
```
[the remark this answers is not in the document]
Comment (nikochiko) [reply tensorflow/python/data/experimental/ops/readers.py]: See https://www.tensorflow.org/community/contribute/docs_ref#testable_docstrings for reference.

File: tensorflow/python/data/experimental/ops/readers.py [hunk 2]
```diff
@@ -420,6 +420,28 @@ def make_csv_dataset_v2(
 
   Raises:
     ValueError: If any of the arguments is malformed.
+  Usage Example:
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: The `Usage Example:` comes above the `Args:` section and below the description. Please see [other docs](https://www.tensorflow.org/versions/r2.1/api_docs/python/tf/concat) and maintain consistency.

File: tensorflow/python/data/experimental/ops/readers.py [hunk 3]
```diff
@@ -420,6 +420,28 @@ def make_csv_dataset_v2(
...
+  ```python
+  >>> train_dataset_url = "https://storage.googleapis.com/download.tensorflow.org/data/iris_training.csv"
+  >>> train_dataset_fp = tf.keras.utils.get_file(fname=os.path.basename(train_dataset_url), origin=train_dataset_url)
+  >>>
+  >>> column_names = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']
+  >>> feature_names = column_names[:-1]
+  >>> label_name = column_names[-1]
+  >>> 
+  >>> batch_size = 32
+  >>> train_dataset = tf.data.experimental.make_csv_dataset(
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: Please add an indent here. See the third code block in the example: https://www.tensorflow.org/tutorials/load_data/csv#load_data

File: tensorflow/python/data/experimental/ops/readers.py [hunk 4]
```diff
@@ -420,6 +420,28 @@ def make_csv_dataset_v2(
...
+  ... train_dataset_fp,
+  ... batch_size,
+  ... column_names=column_names,
+  ... label_name=label_name,
+  ... num_epochs=1)
+  >>> features, labels = next(iter(train_dataset))
+  >>> print(features)
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: If you remove the triple backticks, you will need to provide the expected output. Test your function on Google Colab and see the output. For more information on how to add output to test, see other examples or go through the documentation: https://www.tensorflow.org/community/contribute/docs_ref#docstring_considerations

File: tensorflow/python/data/experimental/ops/readers.py [hunk 5]
```diff
@@ -420,6 +420,28 @@ def make_csv_dataset_v2(
...
+  >>> label_name = column_names[-1]
+  >>> 
+  >>> batch_size = 32
+  >>> train_dataset = tf.data.experimental.make_csv_dataset(
+  ... train_dataset_fp,
+  ... batch_size,
+  ... column_names=column_names,
+  ... label_name=label_name,
+  ... num_epochs=1)
+  >>> features, labels = next(iter(train_dataset))
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: Can you rename it to `features_batch`, `labels_batch`? The name `features` suggests that it is the complete dataset.

File: tensorflow/python/data/experimental/ops/readers.py [hunk 6]
```diff
@@ -420,6 +420,28 @@ def make_csv_dataset_v2(
...
+  >>> train_dataset_url = "https://storage.googleapis.com/download.tensorflow.org/data/iris_training.csv"
+  >>> train_dataset_fp = tf.keras.utils.get_file(fname=os.path.basename(train_dataset_url), origin=train_dataset_url)
+  >>>
+  >>> column_names = ['sepal_length', 'sepal_width', 'petal_length', 'petal_width', 'species']
+  >>> feature_names = column_names[:-1]
+  >>> label_name = column_names[-1]
+  >>> 
+  >>> batch_size = 32
+  >>> train_dataset = tf.data.experimental.make_csv_dataset(
+  ... train_dataset_fp,
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: The code would be easier to read if you add an indent here like in the third codeblock in this section: https://www.tensorflow.org/tutorials/load_data/csv#load_data

Comment (boronhub) [reply tensorflow/python/data/experimental/ops/readers.py]: @nikochiko done

File: tensorflow/python/data/experimental/ops/readers.py [hunk 7]
```diff
@@ -340,6 +340,41 @@ def make_csv_dataset_v2(
...
+  ... batch_size,
+  ... column_names=column_names,
+  ... label_name=label_name,
+  ... num_epochs=1)
+  >>> features_batch, labels_batch = next(iter(train_dataset))
+  >>> print(features_batch)
+  <OrderedDict([('sepal_length', <tf.Tensor: id=138, shape=(32,), dtype=float32, numpy=
+array([4.9, 5.1, 5.4, 6.8, 6.6, 6.7, 5. , 6.2, 5.1, 6.2, 6. , 6.4, 6.2,
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: This is supposed to be aligned similar to the line above. Add necessary whitespace for this and all similar lines.

File: tensorflow/python/data/experimental/ops/readers.py [hunk 8]
```diff
@@ -340,6 +340,41 @@ def make_csv_dataset_v2(
...
+  >>> batch_size = 32
+  >>> train_dataset = tf.data.experimental.make_csv_dataset(
+  ... train_dataset_fp,
+  ... batch_size,
+  ... column_names=column_names,
+  ... label_name=label_name,
+  ... num_epochs=1)
+  >>> features_batch, labels_batch = next(iter(train_dataset))
+  >>> print(features_batch)
+  <OrderedDict([('sepal_length', <tf.Tensor: id=138, shape=(32,), dtype=float32, numpy=
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: The value of `id` may change. Go through this page carefully: https://www.tensorflow.org/community/contribute/docs_ref and accommodate this. 
Also, the value of each Tensor may change. This is the value of `numpy` here. Make the exact value of the `array` variable, keeping the `dtype` as `float32`.

**Review by nikochiko — CHANGES_REQUESTED** [review]

**Comment by nikochiko** [comment]
@boronhub Few changes. Rest LGTM.

**Comment by boronhub** [comment]
please check now

File: tensorflow/python/data/experimental/ops/readers.py [hunk 9]
```diff
@@ -340,6 +340,41 @@ def make_csv_dataset_v2(
...
+  >>> batch_size = 32
+  >>> train_dataset = tf.data.experimental.make_csv_dataset(
+  ...   train_dataset_fp,
+  ...   batch_size,
+  ...   column_names=column_names,
+  ...   label_name=label_name,
+  ...   num_epochs=1)
+  >>> features_batch, labels_batch = next(iter(train_dataset))
+  >>> print(features_batch)
+  <OrderedDict([('sepal_length', <tf.Tensor: shape=(32,), dtype=float32, numpy=
```
Comment (nikochiko) [inline tensorflow/python/data/experimental/ops/readers.py]: @boronhub We don't want to delete the `id` field. Please go through this documentation: https://www.tensorflow.org/community/contribute/docs_ref
The correct syntax for values we are not sure about is `...`. 
As in the example inside the documentation, you can see `id=...`. Do the same thing here. For all four. Also, the order of values and which may come up is random. So do the same for those. 
That is, 
```
numpy=array([..., dtype=float32])
```
Make the same change for all four cases.

**Review by nikochiko — CHANGES_REQUESTED** [review]
These examples are checked with doctest. So any inconsistencies will cause errors. If you think the examples syntax is confusing, refer the docs page mentioned.

File: tensorflow/python/data/experimental/ops/readers.py (hunk 9 above)
Comment (rachellim) [reply tensorflow/python/data/experimental/ops/readers.py]: +1

**Review by rachellim — COMMENTED** [review]
Thanks for the change! This looks good to me. For consistency with the rest of `readers.py`, I'd prefer the backtick notation (` ``` `) over `>>>`.

**Comment by boronhub** [comment]
Wait, so I remove the >>> notation everywhere, and just use 
```python
<indented code here without `>>>` >
```

**Comment by rachellim** [comment]
I mean in the style of the other example code in the file, [for example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/data/experimental/ops/readers.py#L627-L638). Thanks!

**Comment by boronhub** [comment]
Will the output be preceded by >> ?

File: tensorflow/python/data/experimental/ops/readers.py [hunk 10]
```diff
@@ -340,7 +340,32 @@ def make_csv_dataset_v2(
   tuple that corresponds to a batch of CSV rows. The features dictionary
   maps feature column names to `Tensor`s containing the corresponding
   feature data, and labels is a `Tensor` containing the batch's label data.
-
+  
+  Usage Example:
+  
+  Using IRIS dataset to show how to convert .csv file into a dataset.
+  
+    ```python
```
Comment (rachellim) [inline tensorflow/python/data/experimental/ops/readers.py]: this code block should be at the same indentation level as line 346

File: tensorflow/python/data/experimental/ops/readers.py [hunk 11]
```diff
@@ -340,7 +340,32 @@ def make_csv_dataset_v2(
...
+    train_dataset = tf.data.experimental.make_csv_dataset(
+      train_dataset_fp,
+      batch_size,
+      column_names=column_names,
+      label_name=label_name,
+      num_epochs=1)
+    features_batch, labels_batch = next(iter(train_dataset))
+    print(features_batch)
+    
+    <OrderedDict([(..., <tf.Tensor: shape=(32,), dtype=float32, numpy=array([...], dtype=float32)>)])>
```
Comment (rachellim) [inline tensorflow/python/data/experimental/ops/readers.py]: Use `>>` for this output line

Comment (boronhub) [reply tensorflow/python/data/experimental/ops/readers.py]: @rachellim ping

**Review by rachellim — DISMISSED** [review]
Thanks!

**Comment by gbaned** [comment]
@boronhub Could you please address Ubuntu Sanity errors? Thanks!

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Please trim trailing whitespace

File: tensorflow/python/data/experimental/ops/readers.py [hunk 12]
```diff
@@ -340,7 +340,28 @@ def make_csv_dataset_v2(
   tuple that corresponds to a batch of CSV rows. The features dictionary
   maps feature column names to `Tensor`s containing the corresponding
   feature data, and labels is a `Tensor` containing the batch's label data.
-
+
```
Comment (mihaimaruseac) [inline tensorflow/python/data/experimental/ops/readers.py]: This line has whitespace at the end of it, this one should have been trimmed.

You can configure your editor to remove trailing whitespace on save or at least to show it to you

File: tensorflow/python/data/experimental/ops/readers.py [hunk 13]
```diff
@@ -340,7 +340,28 @@ def make_csv_dataset_v2(
   tuple that corresponds to a batch of CSV rows. The features dictionary
   maps feature column names to `Tensor`s containing the corresponding
   feature data, and labels is a `Tensor` containing the batch's label data.
-
+  
+  Usage Example:
+  
+  Using IRIS dataset to show how to convert .csv file into a dataset.
+  
+  ```python
```
Comment (mihaimaruseac) [inline tensorflow/python/data/experimental/ops/readers.py]: I'd recommend converting this to a doctest, see guidelines at https://www.tensorflow.org/community/contribute/docs_ref#testable_docstrings

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/experimental/ops/readers.py [hunk 14]
```diff
@@ -362,10 +360,8 @@ def make_csv_dataset_v2(
     num_epochs=1)
   features_batch, labels_batch = next(iter(train_dataset))
   print(features_batch)
-
   >> <OrderedDict([(..., <tf.Tensor: shape=(32,), dtype=float32, numpy=array([...], dtype=float32)>)])>
   ```
-
```
Comment (mihaimaruseac) [inline tensorflow/python/data/experimental/ops/readers.py]: This line should not have been removed

File: tensorflow/python/data/experimental/ops/readers.py (hunk 13 above)
Comment (rachellim) [reply tensorflow/python/data/experimental/ops/readers.py]: Ah, I wasn't aware that `>>>` == doctest earlier, and requested for the author to use backtick notation for consistency with the rest of the file. Apologies for the confusion.

**Review by mihaimaruseac — DISMISSED** [review]
Although you haven't removed trailing whitespace, approving and hoping internal code formatting will remove it without causing import issues

**Review by rachellim — DISMISSED** [review]

File: tensorflow/python/data/experimental/ops/readers.py [hunk 15]
```diff
@@ -341,6 +341,26 @@ def make_csv_dataset_v2(
   maps feature column names to `Tensor`s containing the corresponding
   feature data, and labels is a `Tensor` containing the batch's label data.
 
+  Usage Example:
+
+  Using IRIS dataset to show how to convert .csv file into a dataset.
+
+  >>> train_dataset_url = "https://storage.googleapis.com/download.tensorflow.org/data/iris_training.csv"
+  >>> train_dataset_fp = tf.keras.utils.get_file(fname=os.path.basename(train_dataset_url), origin=train_dataset_url)
```
Comment (mihaimaruseac) [inline tensorflow/python/data/experimental/ops/readers.py]: Can you rewrite this test so it doesn't hit the network on every test?

This causes issues on test environments with limited/no internet access

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/experimental/ops/readers.py (hunk 15 above)
Comment (boronhub) [reply tensorflow/python/data/experimental/ops/readers.py]: Should I do something like 

`train_dataset_url = "path/to/file.csv"`

Comment (mihaimaruseac) [reply tensorflow/python/data/experimental/ops/readers.py]: Yes, and add the csv to `data` in the corresponding `BUILD` bazel target

Comment (boronhub) [reply tensorflow/python/data/experimental/ops/readers.py]: It will hit the network anyway, can i make it a usage example like @rachellim said ?

Comment (mihaimaruseac) [reply tensorflow/python/data/experimental/ops/readers.py]: Ok

**Review by rachellim — DISMISSED** [review]
Thanks!

**Review by mihaimaruseac — APPROVED** [review]

**Comment by gbaned** [comment]
@boronhub Could you please help fix Ubuntu Sanity failures? Thanks!

**Review by mihaimaruseac — APPROVED** [review]

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Sanity still fails

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]
Pylint on sanity build still fails

File: tensorflow/python/data/experimental/ops/readers.py [hunk 16]
```diff
@@ -342,25 +342,24 @@ def make_csv_dataset_v2(
...
+  batch_size = 32
+  train_dataset = tf.data.experimental.make_csv_dataset(
+    train_dataset_fp,
+    batch_size,
+    column_names=column_names,
+    label_name=label_name,
+    num_epochs=1)
+  features_batch, labels_batch = next(iter(train_dataset))
+  print(features_batch)
+  >> <OrderedDict([(..., <tf.Tensor: shape=(32,), dtype=float32, numpy=array([...], dtype=float32)>)])>
```
Comment (mihaimaruseac) [inline tensorflow/python/data/experimental/ops/readers.py]: Why this change?

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

File: tensorflow/python/data/experimental/ops/readers.py (hunk 16 above)
Comment (boronhub) [reply tensorflow/python/data/experimental/ops/readers.py]: For consistency with the rest of the file.

Comment (mihaimaruseac) [reply tensorflow/python/data/experimental/ops/readers.py]: We would rather have https://github.com/tensorflow/tensorflow/blob/fda0768976e19aec6f7cb7b0c0a54ec160243a93/tensorflow/python/data/experimental/ops/readers.py#L663-L669 and other similar examples (although this one is the only one that has `>>` on the output) rewritten to the doctest format.

Plus, this doesn't solve the sanity pylint failure at all.

**Review by mihaimaruseac — CHANGES_REQUESTED** [review]

**Comment by boronhub** [comment]
Do I lint using regular pylint, or should I use the TensorFlow config file ?

**Comment by mihaimaruseac** [comment]
You should use TensorFlow's config.

File: tensorflow/python/data/experimental/ops/readers.py [hunk 17]
```diff
@@ -341,6 +341,34 @@ def make_csv_dataset_v2(
...
+  batch_size = 32
+  train_dataset = tf.data.experimental.make_csv_dataset(
+    train_dataset_fp,
+    batch_size,
+    column_names=column_names,
+    label_name=label_name,
+    num_epochs=1)
+  features_batch, labels_batch = next(iter(train_dataset))
+  print(features_batch)
+  >> <OrderedDict([(..., <tf.Tensor: shape=(32,), dtype=float32, numpy=
```
Comment (mihaimaruseac) [inline tensorflow/python/data/experimental/ops/readers.py]: At least replace `>>` with `#` to make this block look like valid code.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by boronhub** [comment]
@mihaimaruseac builds passing

**Review by mihaimaruseac — APPROVED** [review]

**Comment by mihaimaruseac** [comment]
It should get imported shortly.

**Comment by boronhub** [comment]
@mihaimaruseac ETA on merge ?

**Comment by boronhub** [comment]
@mihaimaruseac ETA on merge ?

**Comment by mihaimaruseac** [comment]
Sorry for the delay, just out of a 3-days weekend.

**Comment by mihaimaruseac** [comment]
Can you try rebasing on master please?

**Comment by boronhub** [comment]
I tried, but there are a lot of conflict errors in files I haven't even edited.

**Comment by mihaimaruseac** [comment]
Probably you started from a bad branch.

You can fix this by:

1. Download https://patch-diff.githubusercontent.com/raw/tensorflow/tensorflow/pull/34943.diff (add `.diff` to the end of the PR URL)
2. Close this PR, delete the branch
3. `git checkout master`; `git pull --rebase`; `git checkout -b your_new_branch`
4. `git am /path/to/downloaded/diff`
5. Make a new PR.

**Comment by gbaned** [comment]
@boronhub Can you please check mihaimaruseac's comments and keep us posted? Thanks!

**Comment by tensorflowbutler** [comment]
It has been 16 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by gbaned** [comment]
I'm going to go ahead and close this PR, because it seems to have stalled. If you're still interested in pursing this (and responding to my comments), please feel free to reopen!
